src/mlir/design.md

```diff
@@ -0,0 +1,158 @@
+# Verona MLIR Generator Design Choices
+
+The MLIR generator sits in between the AST and LLVM IR, and is meant to ease
+lowering while still maintaining some high level concepts to allow us to optimise
+the code without having to scan the whole IR on every pass.
+
+The LLVM pipeline intermixes analysis passes with transformation passes and has
+a way of tracking which transformation pass invalidates which analysis pass to
+avoid re-running complex analysis passes. However, more often than not, the
+transformation passes end up invalidating them anyway and a new run is needed.
+
+In Verona, we'll be bundling all modules, including the standard library and all
```

Comment:
I think there are two things being conflated here:

 - The compilation model requiring whole-program reachability analysis for reification.
 - A design choice to emit everything in a single MLIR module.

Long term, I'd hope that we'd have an internal compiler structure that is closer to thin-LTO: we need to walk the entire call graph for reachability, but a lot of the rest of the pipeline is amenable to parallelisation.  I'd like us to eventually end up with a load of different LLVM IR modules and module summary analyses that we can invalidate and regenerate independently when the corresponding Verona code changes and run the codegen pipeline on in parallel (for incremental debugging builds, doing less cross-module analysis so that we can keep .o files for individual modules around if we haven't changed any of the things that they depend on).

Reply:
There's also an infrastructure issue. Emitting LLLVM IR form multiple MLIR modules isn't supported (we'd need multiple LLVM modules). So, if we want to keep the MLIR in modules all the way to LTO, we'll have to create multiple LLVM modules and handle them on our side when we link all LLVM IR modules together. This isn't a trivial infrastructure.

So, at the moment, the design decision is to simplify the infrastructure of some things, while we get other things done. Right now, I need to get something from Verona into LLVM IR and this is the simplest route. Simple enough, that adding functionality to it is easier than if we go down a route we "think is what we want" and then we discover it wasn't, and have to refactor everything from one complicated infrastructure into another completely different complicated infrastructure.

Reply:
It would be good to capture this in the document, specifically the split between things that are:

- Intrinsic to the language (e.g. whole-program reachability for reification)
- Shortcuts to make the implementation proceed faster (single MLIR module)
- Long-term goals that we should try to avoid making harder but which won't be in the first implementation (multiple MLIR / LLVM IR modules).

Reply:
I think you're looking for a larger document, one that spans across the whole compiler. This document is restricted to the `mlir` directory because it's only concerned with the MLIR lowering, from the AST. For example, reification and type checks will have already been done by the AST, so it should not be in this document.

The MLIR module discussion is an open problem. I don't really know if it's better or worse to do it that way. I'm trying different things, and those things change with time because MLIR is also changing fast. Currently, the symbol table doesn't seem to work like I expected, so it's quite possible that we'll have to go with single module and link the Verona modules with attributes. But it's too early to tell.

It seems to me like you're expecting a document that describes the whole process, from beginning to end, implemented in a way that is reasonably deterministic and based on well tested methods and infrastructures. We have none of that here.